### ISSUE TITLE
ca65: Corrupted object files when compiling source files with Macintosh line endings (CR)

### ISSUE THREAD

**Issue by Gumball2415** [cc65/cc65#1894]
Macintosh style line endings (CR) in a source assembly file seem to produce incomplete/corrupt object files.
Here is a short demo program:
[hello_NES.zip](https://github.com/cc65/cc65/files/9912587/hello_NES.zip)

and here are the commands used to compile the program:
```
ca65 -g -o hello_NES.o hello_NES.asm
ld65 -v -t nes --dbgfile hello_NES.dbg hello_NES.o -o hello_NES.nes
```
I'm using cc65 version at commit 911ff40f6, built with MinGW 64-bit MSYS2.

When the source assembly file contains only CR line endings, this is the resulting object file:
[hello_NES_cr.zip](https://github.com/cc65/cc65/files/9912564/hello_NES_cr.zip)

And correspondingly, here is the resulting object file with only LF line endings:
[hello_NES_lf.zip](https://github.com/cc65/cc65/files/9912567/hello_NES_lf.zip)

**Comment by bbbradsmith** [comment]
So, what happens in this case is the first line starts with ```;``` beginning a comment. The [scanner's comment loop](https://github.com/cc65/cc65/blob/805e98a7aa6fbd2b419de5192163f7d47c1f3beb/src/ca65/scanner.c#L1357) discards characters until '\n' or EOF. Because your file has no '\n' the comment will last until the end of the file, producing no errors and an empty object.

So, that's the problem. I don't have any quick suggestion about how to address it. I'm a little curious how these line endings came about... I haven't personally had to deal with this kind of line endings in a very long time, and I haven't spotted any of my current Mac tools doing it. Is there a place this still comes up in normal practice?

**Comment by bbbradsmith** [comment]
As far as I can tell, ca65's scanner doesn't really look for anything but '\n' to mark a new line. It [uses "r" with fopen](https://github.com/cc65/cc65/blob/805e98a7aa6fbd2b419de5192163f7d47c1f3beb/src/ca65/scanner.c#L504) and reads the file [with fgetc](https://github.com/cc65/cc65/blob/805e98a7aa6fbd2b419de5192163f7d47c1f3beb/src/ca65/scanner.c#L386).

So, on Windows opening a file with Windows endings, the '\r' will get hidden from the program anyway, and it will only see the '\n' for line endings. I assume if you opened a Windows file on Linux, the '\r' would show up. However '\r' is [treated as whitespace](https://github.com/cc65/cc65/blob/5d84a4ba137e32163390ccd058c32ad169b4ca94/src/common/chartype.c#L89), so it might not make a significant difference to the assembly either way. Maybe some extra bytes in the debug info?

I would expect that if it were run on an older Mac OS, fopen/fgetc would also replace it's '\r' line ending patterns with '\n' and it should parse correctly? Though, I'd also expect current Mac OS not to do that...

Anyway, maybe the OS differences aren't so important. I think we could adjust [IFNextChar](https://github.com/cc65/cc65/blob/805e98a7aa6fbd2b419de5192163f7d47c1f3beb/src/ca65/scanner.c#L376) to look for both '\n' and '\r', and after detecting an '\r' make sure to skip an '\n' if it's the next character read?

**Comment by bbbradsmith** [comment]
I created #2078 because it seemed like it could be done with a pretty small change.

I don't know to what extent this line ending issue is a problem, so I'm not for or against the PR being merged, but please comment if you have an opinion.

**Comment by mrdudz** [comment]
fixed via #2078
